src/modules/kvs/kvs.c

```diff
@@ -203,37 +203,59 @@ static int event_subscribe (kvs_ctx_t *ctx, const char *ns)
             goto cleanup;
         }
 
+        /* On rank 0, we need to listen for all of these namespace
+         * events, all of the time.  So subscribe to them just once on
+         * rank 0. */
+        if (ctx->rank == 0) {
+            if (flux_event_subscribe (ctx->h, "kvs.setroot") < 0) {
+                flux_log_error (ctx->h, "flux_event_subscribe");
+                goto cleanup;
+            }
+
+            if (flux_event_subscribe (ctx->h, "kvs.error") < 0) {
+                flux_log_error (ctx->h, "flux_event_subscribe");
+                goto cleanup;
+            }
+
+            if (flux_event_subscribe (ctx->h, "kvs.namespace-remove") < 0) {
+                flux_log_error (ctx->h, "flux_event_subscribe");
+                goto cleanup;
+            }
+        }
+
         ctx->events_init = true;
     }
 
-    if (asprintf (&setroot_topic, "kvs.setroot-%s", ns) < 0) {
-        errno = ENOMEM;
-        goto cleanup;
-    }
+    if (ctx->rank != 0) {
+        if (asprintf (&setroot_topic, "kvs.setroot-%s", ns) < 0) {
+            errno = ENOMEM;
```

Comment:
`asprintf()` already sets errno on error, so might want to drop the explicit `errno = ENOMEM` assignment in the cases you are touching in this PR.

Since the missing coverage is mainly these error paths, it might increase coverage a bit too.